### PR TITLE
H2SQL persistence service

### PR DESCRIPTION
**Pull request by cdjackson**
Initial commit of the H2 persistence service that implements the new persistence interfaces.

Note - as discussed on the previous thread, I'm using the current 'stable' version of H2 which is far from the most recent (about 2 years old!). I'm happy to use another version if that's desirable, but figured 'stable' was best! I understand there is a CQ already done for a more recent version, so it might make sense to use that - up to you...

### CODE REVIEW

**Comment by lewie** [comment]
@cdjackson, why not using JDBC Service?

**Comment by cdjackson** [comment]
This will provide a simple service for ESH inclusion so that there’s a reference implementation.

**Comment by lewie** [comment]
Ah cool! :-)

**Comment by cdjackson** [comment]
I should add, I did consider porting over the JDBC binding, but I think all the licensing would make it a nightmare under ESH so I suspect it might be better to port that into OH2…

One other point - I did some profiling on this to check memory size, and speed and ended up using strings for most data except numbers as this was significantly small file size (in H2 at least).

**Comment by lewie** [comment]
There was a discussion after I had written a postgreSQL Bundle. I have started JDBC bundle to bring together all the redundant, similar and duplicated database code (mySQL, your H2, my posgreSQL, needed sqLite). To give them a common code base.

The old discussion, code you don't write, you don't have to maintain.

Well, it's like that. ;-)

**Comment by cdjackson** [comment]
> There was a discussion after I had written a postgreSQL Bundle. I have started JDBC bundle to bring together all the redundant, similar and duplicated database code (mySQL, your H2, my posgreSQL, needed sqLite). To give them a common code base.
> 
Sure - I recall this. I’m not sure of your point though, sorry.  As I said earlier when you asked why not use the JDBC, I did consider to do this, but I think for ESH where licensing restrictions are considerably more limited than OH, it was better to produce this reference implementation with a single database.

File: extensions/persistence/org.eclipse.smarthome.persistence.h2sql/pom.xml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	xmlns="http://maven.apache.org/POM/4.0.0"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+	<parent>
+		<groupId>org.eclipse.smarthome.persistence</groupId>
+		<artifactId>pom</artifactId>
+		<version>0.9.0-SNAPSHOT</version>
+	</parent>
+
+	<name>H2SQL Persistence Service</name>
```
Comment (maggu2810) [inline extensions/persistence/org.eclipse.smarthome.persistence.h2sql/pom.xml]: Add "Eclipse SmartHome" in front of this (see e.g. #1997)

Comment (cdjackson) [inline extensions/persistence/org.eclipse.smarthome.persistence.h2sql/pom.xml]: Done

**Comment by maggu2810** [comment]
@kaikreuzer Should I try to create a CQ for "h2 Database Version: 1.3.176"?

**Comment by kaikreuzer** [comment]
@maggu2810 Yes, please!

**Comment by kaikreuzer** [comment]
Can this be closed in favor of #2081?

**Comment by cdjackson** [comment]
I've not looked at the other PR at all, but if it covers this PR and I assume doesn't change the implementation, just tidies up the h2 libraries, then it's fine by me.

**Comment by maggu2810** [comment]
It changes your implementation a little bit.
For example: The code base of this PR is not able to store an StringItem's state that contains a quote character. Also not using non-prepared statements will be very insecure against SQL injections etc.

**Comment by maggu2810** [comment]
Also I started to use try-with-resources to prevent resource leaks by missing closes etc.

**Comment by cdjackson** [comment]
Ok  I also started adding prepared statements as well, but since I wasn't sure if this PR was wanted as no-one commented on the PR, I never pushed the update!

I assume it doesnt change the database structure does it?

**Comment by kaikreuzer** [comment]
> but since I wasn't sure if this PR was wanted as no-one commented on the PR

Hey, you should be fully aware that all your contributions are always welcome and that "no response" is usually only due to lack of time!

**Comment by maggu2810** [comment]
> I assume it doesnt change the database structure does it?

ATM I have not done any database structure changes. But as it is not merged yet, I don't think that we must not change the structure.
I changed the naming from "h2sql" to "h2" because this is IIRC the official name "h2 database" (not h2 sql database).
I also changed the code to store UndefType states to the database to, because it is valid that a channel updates it state to e.g UnDefType.UNDEF.

**Comment by cdjackson** [comment]
What are you planning to change in the database structure? I think it should be avoided if possible.

Even if it's not merged, that doesn't mean no-one is using it (I am for starters) so if you change the structure it will need to be resolved! I have over 1 year of data in this format and I don't say that this isn't also used by others...

**Comment by maggu2810** [comment]
> What are you planning to change in the database structure?

Nothing until now, but I think it is a little bit wired to add a new persistence implementation to a project and expect that no one changes the structure.

Sure, I understand your use cases and I don't think we change the structure just to make someone angry.
But for a new persistence implementation that is going to be part of the ESH project this should be not the first goal - perhaps the second or third

**Comment by cdjackson** [comment]
> but I think it is a little bit wired to add a new persistence implementation to a project and expect that no one changes the structure.

Sorry, but I do think that the structure should remain static - at least once this is merged. If it changes, there should be some level of backaward compatibility, or a migration task. People store data in databases for a long time - you need to consider that.

The persistence in ESH is pretty simple and there shouldn't reall be a reason to change the schema much, but I really think this needs to be a major consideration - sorry! Changes should only be made with good reason.

**Comment by maggu2810** [comment]
Full ack @cdjackson for all that is persistent using the ESH h2 bundle - it should be a must.
For all stuff that has been persistent using another bundle (the pre-ESH h2 bundle) I think it should be done if possible (and I think it will be possible if we keep the schema for the first version), but that "must" is not so strict as the first one.

For other systems I created a table "metainfo" in the database (or something similar) to place meta information for the database itself (current schema version). So it has been pretty easy to check that version after establish the database connection and do the migration to the next version (do it iterative as long as the db version is lower then the impl. version).

**Comment by cdjackson** [comment]
Sounds good - thanks.

**Comment by kirantpatil** [comment]
@cdjackson, any updates if H2 DB as persistence available in ESH ?

**Comment by cdjackson** [comment]
I'm not sure what happened to this PR. It was picked up by @maggu2810 but has been closed. I'm happy to submit my original PR again if people want, but I don't know if this was rejected for any specific reason.

**Comment by kaikreuzer** [comment]
See here: https://github.com/eclipse/smarthome/pull/2081#issuecomment-304440897

So if anybody wants to work on it again, @maggu2810's branch should be revived. Afair, the main issue to be solved about it was that we needed a decision on how to deal with UNDEF types.

**Pull request closed without merging**